File: Recursion/ReverseAStack/ReverseAStack.ts
class ReverseAStack {
    /**
     * Problem:
     * Given a stack of integers, reverse it using recursion.
     * Only stack operations are allowed.
     * No loops or extra data structures for the optimal solution.
     */

    /* ============================================================
     * Approach 1: Brute Force (Using Extra Stack)
     *
     * Idea:
     * - Pop elements from original stack
     * - Push them into a temporary stack
     * - This reverses the order
     *
     * Note:
     * - Uses extra stack
     * - Uses loop, hence violates constraints
     * - Included for understanding
     *
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     *
     * Stable: Yes
     * ============================================================
     */
    static reverseStackBrute(stack: number[]): void {
        const temp: number[] = [];

        while (stack.length > 0) {
            temp.push(stack.pop() as number);
        }

        while (temp.length > 0) {
            stack.push(temp.pop() as number);
        }
    }

    /* ============================================================
     * Approach 2: Recursive Reversal (Core Recursive Solution)
     *
     * Idea:
     * - Pop top element
     * - Recursively reverse remaining stack
     * - Insert popped element at bottom using recursion
     *
     * Fully satisfies problem constraints.
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(n)  // recursion stack
     *
     * Stable: Yes
     * ============================================================
     */
    static reverseStackRecursive(stack: number[]): void {
        if (stack.length === 0) return;

        const top = stack.pop() as number;
        this.reverseStackRecursive(stack);
        this.insertAtBottom(stack, top);
    }

    private static insertAtBottom(stack: number[], value: number): void {
        if (stack.length === 0) {
            stack.push(value);
            return;
        }

        const top = stack.pop() as number;
        this.insertAtBottom(stack, value);
        stack.push(top);
    }

    /* ============================================================
     * Approach 3: Optimal Recursive (Cleaner Abstraction)
     *
     * Idea:
     * - Same as Approach 2
     * - Clear separation of concerns
     * - Most expected interview solution
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(n)
     *
     * Stable: Yes
     * ============================================================
     */
    static reverseStackOptimal(stack: number[]): void {
        this.reverse(stack);
    }

    private static reverse(stack: number[]): void {
        if (stack.length <= 1) return;

        const x = stack.pop() as number;
        this.reverse(stack);
        this.pushToBottom(stack, x);
    }

    private static pushToBottom(stack: number[], x: number): void {
        if (stack.length === 0) {
            stack.push(x);
            return;
        }

        const temp = stack.pop() as number;
        this.pushToBottom(stack, x);
        stack.push(temp);
    }
}

/* ============================================================
 * Driver Code
 * ============================================================
 */
const stack1: number[] = [1, 2, 3, 4];
const stack2: number[] = [...stack1];
const stack3: number[] = [...stack1];

ReverseAStack.reverseStackBrute(stack1);
console.log(stack1);

ReverseAStack.reverseStackRecursive(stack2);
console.log(stack2);

ReverseAStack.reverseStackOptimal(stack3);
console.log(stack3);

File: Recursion/ReverseAStack/ReverseAStack.java
import java.util.Stack;

public class ReverseAStack {

    /**
     * Problem:
     * Given a stack of integers, reverse it using recursion.
     * Only stack operations are allowed.
     * No loops or extra data structures for the optimal solution.
     */

    /* ============================================================
     * Approach 1: Brute Force (Using Extra Stack)
     *
     * Idea:
     * - Pop elements from original stack
     * - Push them into a temporary stack
     * - This reverses the order
     *
     * Note:
     * - Uses extra stack and loops
     * - Violates recursion-only constraint
     * - Included for understanding
     *
     * Time Complexity: O(n)
     * Space Complexity: O(n)
     *
     * Stable: Yes
     * ============================================================
     */
    public static void reverseStackBrute(Stack<Integer> stack) {
        Stack<Integer> temp = new Stack<>();

        while (!stack.isEmpty()) {
            temp.push(stack.pop());
        }

        while (!temp.isEmpty()) {
            stack.push(temp.pop());
        }
    }

    /* ============================================================
     * Approach 2: Recursive Reversal (Core Recursive Solution)
     *
     * Idea:
     * - Pop the top element
     * - Recursively reverse remaining stack
     * - Insert popped element at the bottom using recursion
     *
     * Fully satisfies the problem constraints.
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(n)  // recursion stack
     *
     * Stable: Yes
     * ============================================================
     */
    public static void reverseStackRecursive(Stack<Integer> stack) {
        if (stack.isEmpty()) return;

        int top = stack.pop();
        reverseStackRecursive(stack);
        insertAtBottom(stack, top);
    }

    private static void insertAtBottom(Stack<Integer> stack, int value) {
        if (stack.isEmpty()) {
            stack.push(value);
            return;
        }

        int top = stack.pop();
        insertAtBottom(stack, value);
        stack.push(top);
    }

    /* ============================================================
     * Approach 3: Optimal Recursive (Cleaner Abstraction)
     *
     * Idea:
     * - Same logic as Approach 2
     * - Clear separation of reverse and insert responsibilities
     * - Commonly expected interview solution
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(n)
     *
     * Stable: Yes
     * ============================================================
     */
    public static void reverseStackOptimal(Stack<Integer> stack) {
        reverse(stack);
    }

    private static void reverse(Stack<Integer> stack) {
        if (stack.size() <= 1) return;

        int x = stack.pop();
        reverse(stack);
        pushToBottom(stack, x);
    }

    private static void pushToBottom(Stack<Integer> stack, int x) {
        if (stack.isEmpty()) {
            stack.push(x);
            return;
        }

        int temp = stack.pop();
        pushToBottom(stack, x);
        stack.push(temp);
    }

    /* ============================================================
     * Driver Code
     * ============================================================
     */
    public static void main(String[] args) {
        Stack<Integer> stack1 = new Stack<>();
        stack1.push(1);
        stack1.push(2);
        stack1.push(3);
        stack1.push(4);

        Stack<Integer> stack2 = (Stack<Integer>) stack1.clone();
        Stack<Integer> stack3 = (Stack<Integer>) stack1.clone();

        reverseStackBrute(stack1);
        System.out.println(stack1);

        reverseStackRecursive(stack2);
        System.out.println(stack2);

        reverseStackOptimal(stack3);
        System.out.println(stack3);
    }
}
